bee-message/src/payload/transaction/essence.rs

```diff
@@ -0,0 +1,381 @@
+    outputs: Vec<Output>,
+    payload: Option<Payload>,
+}
+
+impl TransactionEssenceBuilder {
+    /// Creates a new `TransactionEssenceBuilder`.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Adds a version number to a `TransactionEssenceBuilder`.
+    pub fn with_version(mut self, version: u8) -> Self {
+        self.version = Some(version);
+        self
+    }
+
+    /// Adds a timestamp to a `TransactionEssenceBuilder`.
+    pub fn with_timestamp(mut self, timestamp: u64) -> Self {
+        self.timestamp = Some(timestamp);
+        self
+    }
+
+    /// Adds an access pledge ID to a `TransactionEssenceBuilder`.
+    pub fn with_access_pledge_id(mut self, access_pledge_id: [u8; PLEDGE_ID_LENGTH]) -> Self {
+        self.access_pledge_id = Some(access_pledge_id);
+        self
+    }
+
+    /// Adds a consensus pledge ID to a `TransactionEssenceBuilder`.
+    pub fn with_consensus_pledge_id(mut self, consensus_pledge_id: [u8; PLEDGE_ID_LENGTH]) -> Self {
+        self.consensus_pledge_id = Some(consensus_pledge_id);
+        self
+    }
+
+    /// Adds inputs to a `TransactionEssenceBuilder`
+    pub fn with_inputs(mut self, inputs: Vec<Input>) -> Self {
+        self.inputs = inputs;
+        self
+    }
+
+    /// Add an input to a `TransactionEssenceBuilder`.
+    pub fn add_input(mut self, input: Input) -> Self {
+        self.inputs.push(input);
+        self
+    }
+
+    /// Add outputs to a `TransactionEssenceBuilder`.
+    pub fn with_outputs(mut self, outputs: Vec<Output>) -> Self {
+        self.outputs = outputs;
+        self
+    }
+
+    /// Add an output to a `TransactionEssenceBuilder`.
+    pub fn add_output(mut self, output: Output) -> Self {
+        self.outputs.push(output);
+        self
+    }
+
+    /// Add a payload to a `TransactionEssenceBuilder`.
+    pub fn with_payload(mut self, payload: Payload) -> Self {
+        self.payload = Some(payload);
+        self
+    }
+
+    /// Finishes a `TransactionEssenceBuilder` into a `TransactionEssence`.
+    pub fn finish(self) -> Result<TransactionEssence, ValidationError> {
+        let version = self.version.ok_or(ValidationError::MissingField("version"))?;
+        let timestamp = self.timestamp.ok_or(ValidationError::MissingField("timestamp"))?;
+        let access_pledge_id = self
+            .access_pledge_id
+            .ok_or(ValidationError::MissingField("access_pledge_id"))?;
+        let consensus_pledge_id = self
+            .consensus_pledge_id
+            .ok_or(ValidationError::MissingField("consensus_pledge_id"))?;
+
+        if !INPUT_OUTPUT_COUNT_RANGE.contains(&self.inputs.len()) {
+            return Err(ValidationError::InvalidInputCount(self.inputs.len()));
+        }
+
+        if !INPUT_OUTPUT_COUNT_RANGE.contains(&self.outputs.len()) {
+            return Err(ValidationError::InvalidOutputCount(self.outputs.len()));
+        }
+
+        if !matches!(self.payload, None | Some(Payload::Indexation(_))) {
```

Comment:
I'd do this with a match where `None | Some(Payload::Indexation(_)` had an empty arm because the negation was a bit confusing at firs.